File: reveal-sc2-opponent-main/stream-overlay-backend/lib/file-lock.js
// @ts-check
/**
 * file-lock.js -- cross-process lock primitive for the data/ directory.
 *
 * Mirrors core/file_lock.py byte-for-byte on the lockfile contract so
 * the Python replay watcher, the Node stream-overlay backend, and the
 * PowerShell live-phase scanner all coordinate on the same lockfiles.
 * See the Python module's top docstring for the full rationale.
 *
 * Lockfile shape (data/.locks/<safe-name>.lock):
 *
 *   { "pid":   <number>,
 *     "host":  "<hostname>",
 *     "lang":  "node" | "python" | "ps",
 *     "platform": "<os>",
 *     "since": <epoch_ms>,
 *     "stamp": "<ISO8601>" }
 *
 * Acquisition uses fs.openSync(path, 'wx') which has the same
 * O_CREAT|O_EXCL semantics as the Python side. Stale detection
 * compares the holder PID against `process.kill(pid, 0)` and the
 * lock age against the configured threshold.
 *
 * Engineering preamble compliance:
 *   - JSDoc + tsc --checkJs strict.
 *   - Function size <= 30 lines, no magic constants.
 *   - Best-effort logging; never throws on release failure.
 *   - Opt-out via SC2TOOLS_DATA_LOCK_ENABLED=0 for emergency rollback.
 *
 * Example:
 *   const { withFileLock } = require('./file-lock');
 *   await withFileLock('data/MyOpponentHistory.json', async () => {
 *     // safe to atomic-write here
 *   });
 */

'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');

const DEFAULT_TIMEOUT_SEC = 30;
const DEFAULT_STALE_AFTER_SEC = 30;
const DEFAULT_BACKOFF_INITIAL_MS = 5;
const DEFAULT_BACKOFF_MAX_MS = 250;

const LOCK_DIR_NAME = '.locks';
const LOCK_SUFFIX = '.lock';
const ENABLE_ENV_VAR = 'SC2TOOLS_DATA_LOCK_ENABLED';
const DISABLE_VALUE = '0';
const LANG_TAG = 'node';

/**
 * Map a target filename to a deterministic lockfile basename. Matches
 * the Python implementation so cross-language coordination works.
 *
 * @param {string} targetPath
 * @returns {string}
 */
function safeLockName(targetPath) {
  let base = path.basename(targetPath);
  for (const strip of ['.bak', '.tmp_restore']) {
    if (base.endsWith(strip)) base = base.slice(0, -strip.length);
  }
  const safe = base.replace(/[^A-Za-z0-9._-]/g, '_');
  return safe + LOCK_SUFFIX;
}

/**
 * Resolve (and create) the lock directory adjacent to the target file.
 *
 * @param {string} targetPath
 * @returns {string}
 */
function resolveLockDir(targetPath) {
  const parent = path.dirname(path.resolve(targetPath));
  const lockDir = path.join(parent, LOCK_DIR_NAME);
  fs.mkdirSync(lockDir, { recursive: true });
  return lockDir;
}

/**
 * Best-effort liveness probe via signal 0. Same semantics as
 * `kill -0 PID` -- doesn't actually deliver a signal, just tests
 * whether the OS would accept one.
 *
 * @param {number} pid
 * @returns {boolean}
 */
function isPidAlive(pid) {
  if (!Number.isInteger(pid) || pid <= 0) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (err) {
    if (err && err.code === 'ESRCH') return false;
    if (err && err.code === 'EPERM') return true;
    return false;
  }
}

/**
 * Parse the lockfile metadata. Returns null on missing / corrupt.
 *
 * @param {string} lockPath
 * @returns {object|null}
 */
function readLockMeta(lockPath) {
  try {
    const raw = fs.readFileSync(lockPath, 'utf8');
    return JSON.parse(raw);
  } catch (_err) {
    return null;
  }
}

/**
 * Build OUR holder metadata. Used both at acquisition time and at
 * release-verify time.
 *
 * @returns {{pid: number, host: string, lang: string, platform: string,
 *            since: number, stamp: string}}
 */
function makeLockMeta() {
  return {
    pid: process.pid,
    host: os.hostname(),
    lang: LANG_TAG,
    platform: process.platform,
    since: Date.now(),
    stamp: new Date().toISOString().replace(/\.\d{3}/, ''),
  };
}

/**
 * Atomic O_EXCL create. Returns true if WE created the file, false
 * when it already exists.
 *
 * @param {string} lockPath
 * @param {object} meta
 * @returns {boolean}
 */
function tryCreateLockfile(lockPath, meta) {
  let fd;
  try {
    fd = fs.openSync(lockPath, 'wx');
  } catch (err) {
    if (err && err.code === 'EEXIST') return false;
    throw err;
  }
  try {
    const payload = Buffer.from(JSON.stringify(meta, null, 2), 'utf8');
    fs.writeSync(fd, payload, 0, payload.length, 0);
    fs.fsyncSync(fd);
  } finally {
    fs.closeSync(fd);
  }
  return true;
}

/**
 * Decide whether the current holder is stale. A null meta (file
 * unreadable) counts as stale.
 *
 * @param {object|null} meta
 * @param {number} staleAfterSec
 * @returns {boolean}
 */
function isStale(meta, staleAfterSec) {
  if (!meta) return true;
  if (!isPidAlive(meta.pid)) return true;
  if (typeof meta.since === 'number') {
    const ageSec = (Date.now() - meta.since) / 1000;
    if (ageSec >= staleAfterSec) return true;
  }
  return false;
}

/**
 * Attempt to remove a stale lockfile. Returns true on success or when
 * the file is already gone; false if the metadata changed under us
 * (another acquirer is now legitimately holding it).
 *
 * @param {string} lockPath
 * @param {object|null} expected
 * @returns {boolean}
 */
function tryStealLock(lockPath, expected) {
  const current = readLockMeta(lockPath);
  if (current === null) return true;
  // `expected === null` means our caller's first read of the holder
  // metadata failed — typically because the holder was mid-write and
  // the file was briefly opened with FileShare.None (PowerShell), or
  // an antivirus held the bytes for a tick. If the second read NOW
  // returns a valid holder, that holder is healthy; we MUST NOT steal.
  // Loop and let the next iteration re-evaluate staleness with a
  // clean read. Without this guard, a sharing-violation transient on
  // a fresh lockfile causes lost updates across processes.
  if (expected === null) return false;
  if (JSON.stringify(current) !== JSON.stringify(expected)) {
    return false;
  }
  try {
    fs.unlinkSync(lockPath);
    return true;
  } catch (err) {
    if (err && err.code === 'ENOENT') return true;
    return false;
  }
}

/**
 * Promise-friendly sleep with exponential backoff capped at
 * DEFAULT_BACKOFF_MAX_MS.
 *
 * @param {number} attempt
 * @returns {Promise<void>}
 */
function backoffSleep(attempt) {
  const delayMs = Math.min(
    DEFAULT_BACKOFF_INITIAL_MS * (2 ** attempt),
    DEFAULT_BACKOFF_MAX_MS,
  );
  return new Promise((resolve) => setTimeout(resolve, delayMs));
}

class FileLockTimeoutError extends Error {
  constructor(message) {
    super(message);
    this.name = 'FileLockTimeoutError';
  }
}

function isDisabled() {
  return process.env[ENABLE_ENV_VAR] === DISABLE_VALUE;
}

/**
 * Release a lock if-and-only-if the on-disk metadata still matches
 * what we wrote. If a stale-steal swapped a different holder in, we
 * stay out of their way.
 *
 * On Windows, another process briefly holding a read handle on our
 * lockfile (liveness check, antivirus indexer) makes DeleteFile fail
 * with a sharing violation -- surfaced to libuv as EBUSY/EACCES/EPERM.
 * Retry a few times with brief sleeps so we don't leak the lockfile;
 * on POSIX this loop is a no-op because unlink succeeds on the first
 * try regardless of open handles.
 *
 * @param {string} lockPath
 * @param {object} ours
 */
function releaseOwned(lockPath, ours) {
  const current = readLockMeta(lockPath);
  if (current === null) return;
  if (current.pid !== ours.pid || current.since !== ours.since) {
    process.stderr.write(
      `[file-lock] release skipped: holder changed under us `
      + `(${current.pid} != ours ${ours.pid})\n`,
    );
    return;
  }
  const TRANSIENT_CODES = new Set(['EBUSY', 'EACCES', 'EPERM']);
  const MAX_ATTEMPTS = 6;
  let lastErr = null;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt += 1) {
    try {
      fs.unlinkSync(lockPath);
      return;
    } catch (err) {
      if (err && err.code === 'ENOENT') return;
      lastErr = err;
      if (!err || !TRANSIENT_CODES.has(err.code)) break;
      const ms = Math.min(5 * (1 << attempt), 80);
      const end = Date.now() + ms;
      while (Date.now() < end) {
        // intentional empty body — sync sleep so callers stay sync
      }
    }
  }
  if (lastErr) {
    process.stderr.write(`[file-lock] release error: ${lastErr.message}\n`);
  }
}

/**
 * Acquire the lock, run `fn`, release. Always releases on the way out
 * even if `fn` throws.
 *
 * @template T
 * @param {string} targetPath
 * @param {() => Promise<T> | T} fn
 * @param {Object} [opts]
 * @param {number} [opts.timeoutSec]
 * @param {number} [opts.staleAfterSec]
 * @returns {Promise<T>}
 */
async function withFileLock(targetPath, fn, opts) {
  if (isDisabled()) {
    return await fn();
  }
  const timeoutSec = (opts && opts.timeoutSec) || DEFAULT_TIMEOUT_SEC;
  const staleAfterSec = (opts && opts.staleAfterSec) || DEFAULT_STALE_AFTER_SEC;
  const lockDir = resolveLockDir(targetPath);
  const lockPath = path.join(lockDir, safeLockName(targetPath));
  const meta = makeLockMeta();
  const deadline = Date.now() + timeoutSec * 1000;
  let attempt = 0;
  let lastSeen = null;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    if (tryCreateLockfile(lockPath, meta)) {
      try {
        return await fn();
      } finally {
        releaseOwned(lockPath, meta);
      }
    }

    const observed = readLockMeta(lockPath);
    if (isStale(observed, staleAfterSec)) {
      tryStealLock(lockPath, observed);
      attempt = 0;
      lastSeen = null;
      continue;
    }

    if (Date.now() >= deadline) {
      const holderPid = observed && observed.pid ? observed.pid : '?';
      throw new FileLockTimeoutError(
        `file-lock: timeout after ${timeoutSec}s waiting on ${lockPath}; `
        + `current holder pid=${holderPid}`,
      );
    }

    if (JSON.stringify(observed) !== JSON.stringify(lastSeen)) {
      attempt = 0;
      lastSeen = observed;
    } else {
      attempt += 1;
    }
    await backoffSleep(attempt);
  }
}

/**
 * Synchronous variant for hot paths that already block (atomic-fs
 * writers run synchronously). Implementation is a tight while-loop
 * with `Atomics.wait`-style sleeps via `setTimeout`-equivalent
 * busy-waiting: we use a small synchronous sleep helper so callers
 * don't need to be async.
 *
 * @template T
 * @param {string} targetPath
 * @param {() => T} fn
 * @param {Object} [opts]
 * @param {number} [opts.timeoutSec]
 * @param {number} [opts.staleAfterSec]
 * @returns {T}
 */
function withFileLockSync(targetPath, fn, opts) {
  if (isDisabled()) {
    return fn();
  }
  // Stage 7: count + measure lock acquisition. Best-effort require()
  // so a partial install doesn't block the lock.
  let metrics = null;
  try { metrics = require('./data_integrity_metrics'); } catch (_e) { /* */ }
  const _basename = path.basename(targetPath);
  const timeoutSec = (opts && opts.timeoutSec) || DEFAULT_TIMEOUT_SEC;
  const staleAfterSec = (opts && opts.staleAfterSec) || DEFAULT_STALE_AFTER_SEC;
  const lockDir = resolveLockDir(targetPath);
  const lockPath = path.join(lockDir, safeLockName(targetPath));
  const meta = makeLockMeta();
  const deadline = Date.now() + timeoutSec * 1000;
  let attempt = 0;
  let lastSeen = null;
  let contended = false;

  // eslint-disable-next-line no-constant-condition
  while (true) {
    if (tryCreateLockfile(lockPath, meta)) {
      if (metrics) {
        metrics.counterInc('lock_acquired', { basename: _basename });
        if (contended) metrics.counterInc('lock_contended', { basename: _basename });
      }
      try {
        return fn();
      } finally {
        releaseOwned(lockPath, meta);
      }
    }
    contended = true;
    const observed = readLockMeta(lockPath);
    if (isStale(observed, staleAfterSec)) {
      tryStealLock(lockPath, observed);
      attempt = 0;
      lastSeen = null;
      continue;
    }
    if (Date.now() >= deadline) {
      if (metrics) metrics.counterInc('lock_timeout', { basename: _basename });
      const holderPid = observed && observed.pid ? observed.pid : '?';
      throw new FileLockTimeoutError(
        `file-lock: timeout after ${timeoutSec}s waiting on ${lockPath}; `
        + `current holder pid=${holderPid}`,
      );
    }
    if (JSON.stringify(observed) !== JSON.stringify(lastSeen)) {
      attempt = 0;
      lastSeen = observed;
    } else {
      attempt += 1;
    }
    sleepSyncMs(Math.min(
      DEFAULT_BACKOFF_INITIAL_MS * (2 ** attempt),
      DEFAULT_BACKOFF_MAX_MS,
    ));
  }
}

/**
 * Block the event loop for `ms` milliseconds. Used only by
 * withFileLockSync; never on a hot Express request path. The atomic-
 * fs writers run on background ticks (Pulse polling, replay webhook
 * persist) so blocking is acceptable -- a 250ms ceiling keeps it from
 * stacking up.
 *
 * @param {number} ms
 */
function sleepSyncMs(ms) {
  const end = Date.now() + ms;
  // Simple busy wait so we don't import any native binding. The
  // backoff caps at 250ms so worst-case CPU usage is bounded.
  while (Date.now() < end) {
    // intentional empty body
  }
}

module.exports = {
  withFileLock,
  withFileLockSync,
  FileLockTimeoutError,
  // Internals exposed for unit tests only.
  _internals: {
    safeLockName,
    resolveLockDir,
    isPidAlive,
    readLockMeta,
    makeLockMeta,
    tryCreateLockfile,
    isStale,
    tryStealLock,
    releaseOwned,
  },
  DEFAULT_TIMEOUT_SEC,
  DEFAULT_STALE_AFTER_SEC,
  ENABLE_ENV_VAR,
};

File: reveal-sc2-opponent-main/core/file_lock.py
"""
file_lock.py -- cross-process lock primitive for the data/ directory.

Why this exists
---------------
Three independent writers touch ``data/MyOpponentHistory.json`` and
its siblings: the PowerShell live-phase scanner, the Python replay
watcher (this package), and the Node.js stream-overlay backend.
Each writer already does a correct atomic write (write-tmp + fsync +
rename) via its own helper -- ``core/atomic_io.py`` here,
``stream-overlay-backend/lib/atomic-fs.js`` on the Node side, and
``Write-FileAtomic`` in ``Reveal-Sc2Opponent.ps1``.

What's missing is **coordination across processes**. Two writers can
each do a textbook atomic write, but if they both load the live file
into memory at roughly the same time, mutate their in-memory copy,
and then race their renames, the second rename clobbers the first
writer's update. The result is the file looking "fine" (last writer
wins, parses cleanly) but with the loser's mutation silently lost.
And in pathological orderings where one writer's ``.tmp`` is being
fsynced while another writer's rename copies the live file to
``.bak``, we end up with a ``.bak`` that captures a half-written
state -- exactly what corrupted ``MyOpponentHistory.json`` and
``meta_database.json`` repeatedly through April 2026.

How
---
A POSIX-style lockfile in ``data/.locks/<safe-name>.lock`` that is
created with ``O_EXCL`` semantics: the OS guarantees the create is
atomic, so only one acquirer wins. Inside the lockfile we record:

    {
      "pid":   <int>,           # holder's process id
      "host":  "<hostname>",    # for cross-machine paranoia
      "lang":  "python|node|ps",
      "since": <epoch_ms>,      # acquisition time, used for stale check
      "stamp": "<ISO8601>"      # human-readable
    }

If acquisition fails because the lockfile already exists, we read its
metadata and decide:

  - holder PID is alive AND lock is younger than ``stale_after_sec``:
    wait with exponential backoff up to ``timeout_sec``.
  - holder PID is dead:
    steal the lock (unlink + retry the create).
  - holder PID is alive AND lock is older than ``stale_after_sec``:
    the holder is alive but stuck (debugger, AV scan). Wait until
    timeout, then steal -- the alternative is deadlock.

The lock is intentionally **per-target-file**, not a global gate, so
unrelated writers don't serialize. The lock name is derived from the
target path with a deterministic safe-character mapping.

Cross-language compatibility
----------------------------
The lockfile contract is documented above and mirrored byte-for-byte
in ``stream-overlay-backend/lib/file-lock.js`` and the PowerShell
``Lock-FileAtomic`` helper. As long as all three writers honour the
same lockfile and the same staleness rules, they coordinate.

Engineering preamble compliance
-------------------------------
* Pure module: file paths, timeouts, and the clock are injectable for tests.
* Type hints on every public function; mypy --strict clean.
* Functions <= 30 lines, no magic constants (all knobs are module
  constants with documented meaning).
* Atomic primitives: ``os.open(O_CREAT | O_EXCL | O_RDWR | O_TRUNC)``
  for acquisition, ``os.fsync`` after write, ``os.unlink`` for release.
* PII safe: opponent names never appear in lockfiles; only PIDs and
  the target filename.

Example
-------
    >>> from core.file_lock import file_lock  # doctest: +SKIP
    >>> with file_lock("data/MyOpponentHistory.json"):  # doctest: +SKIP
    ...     # safe to read-modify-write the target now
    ...     pass
"""
from __future__ import annotations

import contextlib
import errno
import json
import os
import platform
import socket
import sys
import time
from typing import Any, Callable, Dict, Iterator, Optional

# Default knobs. Override per-call where it matters.
DEFAULT_TIMEOUT_SEC = 30.0
DEFAULT_STALE_AFTER_SEC = 30.0
DEFAULT_BACKOFF_INITIAL_MS = 5
DEFAULT_BACKOFF_MAX_MS = 250

LOCK_DIR_NAME = ".locks"
LOCK_SUFFIX = ".lock"
ENABLE_ENV_VAR = "SC2TOOLS_DATA_LOCK_ENABLED"
DISABLE_VALUE = "0"
LANG_TAG = "python"


def _safe_lock_name(target_path: str) -> str:
    """Map a target filename to a deterministic lockfile basename.

    Drops directory components and replaces every character outside
    ``[A-Za-z0-9._-]`` with ``_``. Two different paths can collide
    iff their basenames differ only by such characters; we treat
    that as a feature -- ``MyOpponentHistory.json`` and
    ``MyOpponentHistory.json.bak`` SHOULD share a lock so the .bak
    write doesn't race the live write.

    Example:
        >>> _safe_lock_name("/x/y/MyOpponentHistory.json")
        'MyOpponentHistory.json.lock'
    """
    base = os.path.basename(target_path)
    # Strip the .bak / .tmp_* / .broken-* suffixes so the same
    # logical file shares one lock.
    for strip in (".bak", ".tmp_restore"):
        if base.endswith(strip):
            base = base[: -len(strip)]
    safe = "".join(c if c.isalnum() or c in "._-" else "_" for c in base)
    return safe + LOCK_SUFFIX


def _resolve_lock_dir(target_path: str) -> str:
    """Lock directory sibling-of the data file (creates it if needed)."""
    parent = os.path.dirname(os.path.abspath(target_path))
    lock_dir = os.path.join(parent, LOCK_DIR_NAME)
    os.makedirs(lock_dir, exist_ok=True)
    return lock_dir


def _is_pid_alive(pid: int) -> bool:
    """True if ``pid`` is currently running. Best-effort, never raises."""
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
        return True
    except OSError as exc:
        if exc.errno == errno.ESRCH:
            return False
        if exc.errno == errno.EPERM:
            # We can't signal it (different user / elevated), but it
            # exists. Treat as alive.
            return True
        return False
    except Exception:  # noqa: BLE001
        return False


def _read_lock_meta(lock_path: str) -> Optional[Dict[str, Any]]:
    """Parse the holder metadata; ``None`` on missing / unreadable / bad JSON."""
    try:
        with open(lock_path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (FileNotFoundError, OSError, json.JSONDecodeError):
        return None


def _make_lock_meta() -> Dict[str, Any]:
    return {
        "pid": os.getpid(),
        "host": socket.gethostname(),
        "lang": LANG_TAG,
        "platform": platform.system(),
        "since": int(time.time() * 1000),
        "stamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }


def _try_create_lockfile(lock_path: str, meta: Dict[str, Any]) -> bool:
    """Attempt the O_EXCL create. True on success, False if it already existed."""
    try:
        flags = os.O_CREAT | os.O_EXCL | os.O_RDWR
        if hasattr(os, "O_BINARY"):
            flags |= os.O_BINARY  # type: ignore[attr-defined]
        fd = os.open(lock_path, flags, 0o644)
    except FileExistsError:
        return False
    except OSError as exc:
        if exc.errno == errno.EEXIST:
            return False
        raise
    try:
        payload = json.dumps(meta, indent=2).encode("utf-8")
        os.write(fd, payload)
        os.fsync(fd)
    finally:
        os.close(fd)
    return True


def _is_stale(meta: Optional[Dict[str, Any]], stale_after_sec: float) -> bool:
    """True if the holder is dead OR the lockfile is older than the threshold."""
    if meta is None:
        return True
    pid = meta.get("pid")
    if not isinstance(pid, int) or not _is_pid_alive(pid):
        return True
    since = meta.get("since")
    if isinstance(since, (int, float)):
        age_sec = (time.time() * 1000 - since) / 1000.0
        if age_sec >= stale_after_sec:
            return True
    return False


def _try_steal(lock_path: str, expected: Optional[Dict[str, Any]]) -> bool:
    """Best-effort unlink of a stale lockfile. Idempotent under contention.

    Returns True if WE removed it (or it was already gone). False means
    another acquirer already swapped in fresh metadata that no longer
    matches what we read; back off and re-loop.
    """
    current = _read_lock_meta(lock_path)
    if current is None:
        # Already gone -- a competing acquirer beat us to it.
        return True
    # ``expected is None`` means our caller's first read of the holder
    # metadata failed (typically because the holder was mid-write and
    # the file was briefly inaccessible -- PowerShell uses
    # FileShare.None, and Windows AV scans new files transiently). If
    # the second read NOW returns valid metadata, that holder is
    # healthy; we MUST NOT steal. Without this guard, a sharing-
    # violation transient on a fresh lockfile lets us silently delete
    # a real holder's lock and produce lost updates.
    if expected is None:
        return False
    if current != expected:
        return False
    try:
        os.unlink(lock_path)
        return True
    except FileNotFoundError:
        return True
    except OSError:
        return False


def _backoff_sleep(attempt: int) -> None:
    """Exponential backoff with a hard ceiling -- keeps CPU off the spin lock."""
    delay_ms = min(
        DEFAULT_BACKOFF_INITIAL_MS * (2 ** attempt),
        DEFAULT_BACKOFF_MAX_MS,
    )
    time.sleep(delay_ms / 1000.0)


class FileLockTimeout(TimeoutError):
    """Raised when a lock can't be acquired within ``timeout_sec``."""


def _is_disabled() -> bool:
    return os.environ.get(ENABLE_ENV_VAR, "1") == DISABLE_VALUE


@contextlib.contextmanager
def file_lock(
    target_path: str,
    *,
    timeout_sec: float = DEFAULT_TIMEOUT_SEC,
    stale_after_sec: float = DEFAULT_STALE_AFTER_SEC,
    clock: Callable[[], float] = time.monotonic,
) -> Iterator[None]:
    """Acquire a cross-process lock for ``target_path`` until the with-block exits.

    Args:
        target_path: Absolute or relative path of the file you're about
            to mutate. The lock is named after the basename so all
            writers across processes share the same lock.
        timeout_sec: Hard ceiling on how long to wait. Raises
            :class:`FileLockTimeout` on expiry.
        stale_after_sec: After this many seconds the lock is treated
            as stale (likely a crashed holder) and stolen.
        clock: Injectable monotonic clock for testing.

    Raises:
        FileLockTimeout: when no acquisition succeeds within
            ``timeout_sec``.

    Example:
        >>> with file_lock("data/MyOpponentHistory.json"):  # doctest: +SKIP
        ...     # safe to atomic-write here
        ...     pass
    """
    if _is_disabled():
        # Opt-out for emergency rollback; behaves as a no-op so the
        # callers don't have to branch.
        yield
        return

    lock_dir = _resolve_lock_dir(target_path)
    lock_path = os.path.join(lock_dir, _safe_lock_name(target_path))
    meta = _make_lock_meta()
    deadline = clock() + timeout_sec
    attempt = 0
    last_seen: Optional[Dict[str, Any]] = None

    while True:
        if _try_create_lockfile(lock_path, meta):
            try:
                yield
            finally:
                _release_owned(lock_path, meta)
            return

        # Acquisition failed. Inspect the holder.
        observed = _read_lock_meta(lock_path)
        if _is_stale(observed, stale_after_sec):
            _try_steal(lock_path, observed)
            # Loop and retry the create. No backoff sleep on a stale
            # steal so we get back into the lock as fast as possible.
            attempt = 0
            last_seen = None
            continue

        if clock() >= deadline:
            raise FileLockTimeout(
                f"file_lock: timeout after {timeout_sec:.1f}s waiting on "
                f"{lock_path}; current holder pid={observed.get('pid') if observed else '?'}"
            )

        if observed != last_seen:
            # Holder changed -- reset the backoff so we're fresh.
            attempt = 0
            last_seen = observed
        else:
            attempt += 1
        _backoff_sleep(attempt)


def _release_owned(lock_path: str, meta: Dict[str, Any]) -> None:
    """Unlink the lockfile only if it still carries OUR metadata.

    If we got pre-empted by a stale-steal somewhere else, the file may
    now belong to a different process. Removing it would orphan that
    holder; safer to log + skip.

    On Windows another process briefly holding a read handle for a
    liveness check (or AV indexing) makes ``os.unlink`` raise
    ``PermissionError`` (ERROR_SHARING_VIOLATION). Retry a few times
    with brief sleeps so we don't leak the lockfile -- on POSIX the
    first attempt always succeeds and the loop exits immediately.
    """
    current = _read_lock_meta(lock_path)
    if current is None:
        return
    if current.get("pid") != meta.get("pid") or current.get("since") != meta.get("since"):
        # Someone else owns it now; stay out of their way.
        print(
            f"[file_lock] release skipped: holder changed under us "
            f"({current.get('pid')} != ours {meta.get('pid')})",
            file=sys.stderr,
        )
        return
    max_attempts = 6
    last_exc: Optional[BaseException] = None
    for attempt in range(max_attempts):
        try:
            os.unlink(lock_path)
            return
        except FileNotFoundError:
            return
        except PermissionError as exc:
            last_exc = exc
        except OSError as exc:
            last_exc = exc
            # EBUSY / EACCES / EPERM are the codes we get on Windows
            # when another process is briefly holding a read handle.
            if exc.errno not in (errno.EBUSY, errno.EACCES, errno.EPERM):
                break
        delay_ms = min(5 * (2 ** attempt), 80)
        time.sleep(delay_ms / 1000.0)
    if last_exc is not None:
        print(f"[file_lock] release error: {last_exc}", file=sys.stderr)


__all__ = [
    "file_lock",
    "FileLockTimeout",
    "DEFAULT_TIMEOUT_SEC",
    "DEFAULT_STALE_AFTER_SEC",
    "ENABLE_ENV_VAR",
]
